File: src/main/java/org/example/linkedlist/DuplicateNumber.java
package org.example.linkedlist;

public class DuplicateNumber {
    public int findDuplicate(int[] nums) {
        int slow = 0, fast = 0;
        do {
            slow = nums[slow];
            fast = nums[nums[fast]];
        } while (slow != fast);

        int slow2 = 0;
        do {
            slow = nums[slow];
            slow2 = nums[slow2];
        } while (slow != slow2);

        return slow;
    }

    public static void main(String[] args) {
        //int[] nums = {1,3,4,2,2};
        int[] nums = {3,1,3,4,2};
        System.out.println(new DuplicateNumber().findDuplicate(nums));
    }
}

File: src/main/java/org/example/linkedlist/DuplicateNumber.py
from typing import List


class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow, fast = 0, 0  # we start at 0 because 0 is never part of the cycle, values are between [1, n]
        while True:  # loop to find intersection
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break

        slow2 = 0
        while True:  # loop to find the start of LL cycle
            slow = nums[slow]
            slow2 = nums[slow2]
            if slow == slow2:
                return slow
